lib/ClangImporter/ImportType.cpp

```diff
@@ -2230,6 +2230,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
   return {swiftResultTy, importedType.isImplicitlyUnwrapped()};
 }
 
+static ImportTypeKind
+getImportTypeKindForParam(const clang::ParmVarDecl *param) {
+  ImportTypeKind importKind = ImportTypeKind::Parameter;
+  if (param->hasAttr<clang::CFReturnsRetainedAttr>())
+    importKind = ImportTypeKind::CFRetainedOutParameter;
+  else if (param->hasAttr<clang::CFReturnsNotRetainedAttr>())
+    importKind = ImportTypeKind::CFUnretainedOutParameter;
+
+  return importKind;
+}
+
+Optional<swift::Type> ClangImporter::Implementation::importParameterType(
+    const clang::ParmVarDecl *param, OptionalTypeKind optionalityOfParam,
+    ImportTypeKind importKind, bool allowNSUIntegerAsInt,
+    bool isNSDictionarySubscriptGetter, bool paramIsError,
+    bool paramIsCompletionHandler,
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
```

Comment:
Actual question: do you think there's some way we could get rid of some of the parameters? I think this is a fairly confusing API interface. Maybe we can compute some of these parameters inside of `importParameterType`. Maybe we can also get rid of the last two out parameters by returning a `ParamDecl *`.

Reply:
I think some of them can be simplified, but not all of them:

- `optionalityOfParam`: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods. 
- `importKind` can be a call to `getImportTypeKindForParam`. I didn't do it to avoid more changes, but it can be done now as a NFC commit on top of this.
- `allowNSUIntegerAsInt`: the logic for methods and functions is way too different. Methods depend on selector names.
- `isNSDictionarySubscriptGetter`: only applies to methods. To calculate it one has to pass the `SpecialMethodKind`, which does not apply to functions.
- `paramIsError`: only applies to methods. To calculate one has to pass the `ForeignErrorConvention::Info` (which comes from the `methodName`) and also the `paramIndex`.
- `paramIsCompletionHandler`: similar story. Depends on the `paramIndex` and the `asyncInfo` (that comes from the `methodName`).
- `completionHandlerErrorParamIndex`: I have doubts about this myself. I think this one and the previous one represents different things: one is if the parameter is the completion handler, and the other points to the index of the _error_ of the completion handler. If someone can assure me that they actually represent the same thing, these two can be folded into one.
- `genericParams`: only applies to functions. They are calculated in `ImportDecl.cpp` from the `FuncTemplateDecl`.
- `addImportDiagnosticFn`: passed in order to be passed down to other functions.
- `isInOut` and `isParamTypeImplitlyUnwrapped`: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the `swift::Type` and turning it into a `ParamDecl` there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two `out bool` instead. I can also create a small type with a `swift::Type` and these two `bool` and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.

Reply:
Makes sense, thank you for explaining! 

> optionalityOfParam: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods.

This might be a bug. Why shouldn't it apply to methods?

> genericParams: only applies to functions. They are calculated in ImportDecl.cpp from the FuncTemplateDecl.

Again, seems like it should maybe apply to methods? Or do we not support templated methods?

> isInOut and isParamTypeImplitlyUnwrapped: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the swift::Type and turning it into a ParamDecl there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two out bool instead. I can also create a small type with a swift::Type and these two bool and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.

OK, maybe we can do this change after the fact.

Reply:
> > optionalityOfParam: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods.
> 
> This might be a bug. Why shouldn't it apply to methods?
>
> > genericParams: only applies to functions. They are calculated in ImportDecl.cpp from the FuncTemplateDecl.
> 
> Again, seems like it should maybe apply to methods? Or do we not support templated methods?

Well, Obj-C++ is a strange beast, but I don't think think Obj-C type methods can have C++-like template parameters. They have the lightweight generics, but I don't think you can mix them both. I might be wrong. Also, the language reference is whatever compiles… so no idea how to search for that.

> > isInOut and isParamTypeImplitlyUnwrapped: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the swift::Type and turning it into a ParamDecl there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two out bool instead. I can also create a small type with a swift::Type and these two bool and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.
> 
> OK, maybe we can do this change after the fact.

I can add an extra commit into the stack with this change. I think it would make Puyan idea of mostly NFC commits a little less clear if I do the change in the extraction commit.